libfwupd/fwupd-client.c

```diff
@@ -1089,6 +1073,81 @@ fwupd_client_install (FwupdClient *client,
 		return FALSE;
 	}
 	return TRUE;
+}
+#endif
+
+/**
+ * fwupd_client_install:
+ * @client: A #FwupdClient
+ * @device_id: the device ID
+ * @filename: the filename to install
+ * @install_flags: the #FwupdInstallFlags, e.g. %FWUPD_INSTALL_FLAG_ALLOW_REINSTALL
+ * @cancellable: the #GCancellable, or %NULL
+ * @error: the #GError, or %NULL
+ *
+ * Install a file onto a specific device.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 0.7.0
+ **/
+gboolean
+fwupd_client_install (FwupdClient *client,
+		      const gchar *device_id,
+		      const gchar *filename,
+		      FwupdInstallFlags install_flags,
+		      GCancellable *cancellable,
+		      GError **error)
+{
+#ifdef HAVE_GIO_UNIX
+	FwupdClientPrivate *priv = GET_PRIVATE (client);
+
+	g_return_val_if_fail (FWUPD_IS_CLIENT (client), FALSE);
+	g_return_val_if_fail (device_id != NULL, FALSE);
+	g_return_val_if_fail (filename != NULL, FALSE);
+	g_return_val_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable), FALSE);
+	g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+	/* connect */
+	if (!fwupd_client_connect (client, cancellable, error))
+		return FALSE;
+
+	/* allow ANOTHER_WRITE_REQUIRED */
+	for (guint retries = 0; ; retries++) {
+		g_autoptr(FwupdDevice) device = NULL;
+		g_autoptr(GError) error_local = NULL;
+
+		/* daemon got stuck */
+		if (retries > 5) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_INTERNAL,
+					     "aborting device install loop, limit 5");
+			return FALSE;
+		}
+		if (retries > 0)
+			g_debug ("install try %u", retries + 1);
+		if (!fwupd_client_install_try (client, device_id, filename,
+					       install_flags, cancellable, error))
+			return FALSE;
+
+		/* device not found is perfectly fine */
+		if (priv->device_id_restart == NULL)
+			break;
+		device = fwupd_client_get_device_by_id (client,
+							priv->device_id_restart,
+							cancellable,
+							&error_local);
+		if (device == NULL) {
+			g_debug ("failed to find device: %s", error_local->message);
+			break;
+		}
+		if (!fwupd_device_has_flag (device, FWUPD_DEVICE_FLAG_ANOTHER_INSTALL_REQUIRED))
+			break;
+	}
+
+	/* success */
+	return TRUE;
```

Comment:
I think you're missing a similar loop for fwupdtool too, no?

Reply:
Ya, I wanted some feedback on the general idea before sinking too much time into it :)